pkg/trace/agent/agent.go

```diff
@@ -220,87 +222,205 @@ func (a *Agent) setRootSpanTags(root *pb.Span) {
-			p.RemoveChunk(i)
-			continue
-		}
+func normalizeChunkTF(m *api.Metadata, pt *traceutil.ProcessedTrace) error {
+	normalizeChunk(pt.TraceChunk, pt.Root)
+	return nil
+}
 
-		// Root span is used to carry some trace-level metadata, such as sampling rate and priority.
-		root := traceutil.GetRoot(chunk.Spans)
-		normalizeChunk(chunk, root)
-		if !a.Blacklister.Allows(root) {
-			log.Debugf("Trace rejected by ignore resources rules. root: %v", root)
-			ts.TracesFiltered.Inc()
-			ts.SpansFiltered.Add(tracen)
-			p.RemoveChunk(i)
-			continue
+func blocklist(a *Agent) Transformer {
+	return func(m *api.Metadata, pt *traceutil.ProcessedTrace) error {
+		if !a.Blacklister.Allows(pt.Root) {
+			m.Source.TracesFiltered.Inc()
+			m.Source.SpansFiltered.Add(int64(len(pt.TraceChunk.Spans)))
+			return fmt.Errorf("Trace rejected by ignore resources rules. root: %v", pt.Root)
 		}
+		return nil
+	}
+}
 
-		if filteredByTags(root, a.conf.RequireTags, a.conf.RejectTags) {
-			log.Debugf("Trace rejected as it fails to meet tag requirements. root: %v", root)
-			ts.TracesFiltered.Inc()
-			ts.SpansFiltered.Add(tracen)
-			p.RemoveChunk(i)
-			continue
+func tagFilter(a *Agent) Transformer {
+	return func(m *api.Metadata, pt *traceutil.ProcessedTrace) error {
+		if filteredByTags(pt.Root, a.conf.RequireTags, a.conf.RejectTags) {
+			m.Source.TracesFiltered.Inc()
+			m.Source.SpansFiltered.Add(int64(len(pt.TraceChunk.Spans)))
+			return fmt.Errorf("Trace rejected as it fails to meet tag requirements. root: %v", pt.Root)
 		}
+		return nil
+	}
+}
 
-		// Extra sanitization steps of the trace.
-		for _, span := range chunk.Spans {
-			for k, v := range a.conf.GlobalTags {
-				if k == tagOrigin {
-					chunk.Origin = v
-				} else {
-					traceutil.SetMeta(span, k, v)
+func eachSpan(fs ...SpanTransformer) Transformer {
+	return func(m *api.Metadata, pt *traceutil.ProcessedTrace) error {
+		for _, span := range pt.TraceChunk.Spans {
+			for _, f := range fs {
+				if err := f(m, pt, span); err != nil {
+					return err
 				}
 			}
-			if a.ModifySpan != nil {
-				a.ModifySpan(span)
-			}
-			a.obfuscateSpan(span)
-			Truncate(span)
-			if p.ClientComputedTopLevel {
-				traceutil.UpdateTracerTopLevel(span)
+		}
+		return nil
+	}
+}
+
+func applyGlobalTags(a *Agent) SpanTransformer {
+	return func(m *api.Metadata, pt *traceutil.ProcessedTrace, s *pb.Span) error {
+		for k, v := range a.conf.GlobalTags {
+			if k == tagOrigin {
+				pt.TraceChunk.Origin = v
+			} else {
+				traceutil.SetMeta(s, k, v)
 			}
 		}
-		a.Replacer.Replace(chunk.Spans)
+		return nil
+	}
+}
+
+func modifySpan(a *Agent) SpanTransformer {
+	return func(m *api.Metadata, pt *traceutil.ProcessedTrace, s *pb.Span) error {
+		if a.ModifySpan != nil {
+			a.ModifySpan(s)
+		}
+		return nil
+	}
+}
+
+func obfuscateSpan(a *Agent) SpanTransformer {
+	return func(m *api.Metadata, pt *traceutil.ProcessedTrace, s *pb.Span) error {
+		a.obfuscateSpan(s)
+		return nil
+	}
+}
```

Comment:
Same here. I don't need to return a function that calls an agent method. I can just change the signature on the agent method and use that.